### PR TITLE
Update cached urls in the background

### PR DESCRIPTION
**Pull request by omsmith**
TODO:
- [x] Update README to suggest adding an `'error'` listener

### CODE REVIEW

**Comment by omsmith** [comment]
This is to try to further reduce Hypermedia Proxy's latency in the APAC regions. Don't know that this would be 100% desirable for other applications, but it's probably not problematic for them either...

File: src/index.js [hunk 1]
```diff
@@ -189,9 +195,9 @@ LandlordClient.prototype.lookupTenantUrl = function lookupTenantUri(tenantId) {
 			var url = value.url;
 
 			if (value.expiry <= self._clock()) {
-				return doAndCacheLookup()
-					.catch(function() {
-						return url;
+				doAndCacheLookup()
+					.catch(function(err) {
+						self.emit('error', err);
```
Comment (omsmith) [inline src/index.js]: TODO(omsmith): this will raise the error multiple times for a single deduped background lookup

Comment (omsmith) [reply src/index.js]: Fixed by merging https://github.com/Brightspace/node-landlord-client/pull/17, which dedupes the `lookupTenantUrl` function as a whole

File: README.md [hunk 2]
```diff
@@ -77,6 +78,14 @@ a String.
...
 Checks availability by pinging Landlord. Returns a Promise to a String.
 
+___
+
+#### `.on('error', Function callback)` -> `LandlordClient`
+
+`LandlordClient` is an `EventEmitter` and will emit `error` events for errors
+which occur outside of the normal workflow (such as background fetches). You
+_should_ add an error handler, perhaps just to your logger.
+
```
Comment (jkehres) [inline README.md]: Maybe note that if you don't add an error handler that your process will exit when an error occurs. From the Node.js docs:

> If an `EventEmitter` does not have at least one listener registered for the `'error'` event, and an `'error'` event is emitted, the error is thrown, a stack trace is printed, and the Node.js process exits.

Comment (omsmith) [reply README.md]: Hope people know that, but yeah I'll mention it.

Comment (omsmith) [reply README.md]: Done

**Comment by jkehres** [comment]
This needs a package.json version bump - unless you were planning to handle that in a separate PR.

**Comment by omsmith** [comment]
I always handle the release process separate from changes. It's labelled as semver-major so when reviewing what changes went in, I can decide on the version change.

**Comment by jkehres** [comment]
Is it worth making this async behaviour configurable? In other words, you can opt-in/out of this behaviour when constructing the client. This would allow this to be minor version change instead of a major version change.

**Comment by jkehres** [comment]
Should a similar strategy of async lookups be applied to `lookupTenantId`?

**Comment by omsmith** [comment]
I'd say the reason to make it configurable would be if we think there are situations where we would desire the different behaviours in different applications. Not worried about what version bump it ends up being.

**Comment by omsmith** [comment]
`lookupTenantId` is a permanent cache, this wouldn't apply.

**Comment by omsmith** [comment]
The only situation that I would see as ~potentially problematic is a long-running process (so, probably not a lambda, but no guarantees I suppose) which interacts with a tenant irregularly.

As an example, though the timelines can be extended indefinitely
08:00 lookup URL, receive value A, cache, use value A
11:00 landlord entry is updated to value B
14:00 use value A, lookup URL, receive value B, cache

However if this leads to a failure... I would hope any process would get kicked off to try again.

**Pull request merged**